activation/poet.go

```diff
@@ -118,41 +118,64 @@ func (c *HTTPPoetClient) Submit(ctx context.Context, challenge []byte, signature
 
 // PoetServiceID returns the public key of the PoET proving service.
 func (c *HTTPPoetClient) PoetServiceID(ctx context.Context) (types.PoetServiceID, error) {
-	if c.poetServiceID != nil {
+	if c.poetServiceID.ServiceID != nil {
 		return c.poetServiceID, nil
 	}
-	resBody := rpcapi.GetInfoResponse{}
+	resBody := rpcapi.InfoResponse{}
 
 	if err := c.req(ctx, http.MethodGet, "/v1/info", nil, &resBody); err != nil {
-		return nil, fmt.Errorf("getting poet ID: %w", err)
+		return types.PoetServiceID{}, fmt.Errorf("getting poet ID: %w", err)
 	}
 
-	c.poetServiceID = types.PoetServiceID(resBody.ServicePubkey)
+	c.poetServiceID.ServiceID = resBody.ServicePubkey
 	return c.poetServiceID, nil
 }
 
 // Proof implements PoetProvingServiceClient.
 func (c *HTTPPoetClient) Proof(ctx context.Context, roundID string) (*types.PoetProofMessage, error) {
-	resBody := rpcapi.GetProofResponse{}
+	resBody := rpcapi.ProofResponse{}
 	if err := c.req(ctx, http.MethodGet, fmt.Sprintf("/v1/proofs/%s", roundID), nil, &resBody); err != nil {
 		return nil, fmt.Errorf("getting proof: %w", err)
 	}
 
+	p := resBody.Proof.GetProof()
+	root := p.GetRoot()
+	leaves := make([]shared.Leaf, 0, len(p.ProvenLeaves))
```

Comment:
The original code used getters like `p.GetProvenLeaves()`. The reason is that in protobuf every field is _optional_ and the getters have nil-checks. It is possible that both `err` and `p` are `nil` here - the data from the poet cannot be trusted. Direct access can lead to a segfault.

Reply:
Fixed